File: compiler/bench/jumpsearch.c
// expected: 499500000
// Jump search benchmark - search 1M times in sorted array

#include <stdio.h>

#define SIZE 1000
#define ITERS 1000000
#define STEP 31

int arr[SIZE];

int min(int a, int b) { return a < b ? a : b; }

int jump_search(int val) {
    int prev = 0;
    int curr = STEP;
    while (curr < SIZE && arr[curr - 1] < val) {
        prev = curr;
        curr = min(curr + STEP, SIZE);
    }
    // Linear search from prev to curr
    for (int i = prev; i < min(curr, SIZE); i++) {
        if (arr[i] == val) return i;
    }
    return -1;
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        sum += jump_search(i % SIZE);
    }
    return sum;
}

int main(void) {
    init_arr();
    printf("%ld\n", bench());
    return 0;
}

File: compiler/bench/jumpsearch.fs
\ expected: 499500000
\ Jump search benchmark - search 1M times in sorted array

1000 constant SIZE
1000000 constant ITERS
31 constant STEP
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: jump-search ( val -- idx|-1 )
  0                       \ val prev
  begin
    dup STEP + SIZE min   \ val prev curr
    dup 1- arr@ 2 pick <
    over SIZE < and
  while
    nip                   \ val curr ; prev=curr
  repeat
  \ linear search from prev to min(curr, SIZE)
  SIZE min swap           \ val hi prev
  begin 2dup > while
    dup arr@ 2 pick = if
      nip nip exit
    then
    1+
  repeat
  2drop drop -1 ;

: init-arr ( -- )
  SIZE 0 do i i arr! loop ;

: bench ( -- sum )
  0
  ITERS 0 do
    i SIZE mod jump-search +
  loop ;

: main
  init-arr
  bench . cr ;
